components/modals/FilterModal.tsx

```diff
@@ -1,15 +1,284 @@
-import React from 'react';
-import { View } from 'react-native';
-
-// TODO: filter modal
-/**
- * -  should allow user to set level, topic, rooms, and session type filters
- * -  clicking the filter button should close the modal and set the filters on the sessions page
- * -  clicking cancel should close the modal
- */
-
-const FilterModal = () => {
-  return <View />;
+import { AntDesign } from '@expo/vector-icons';
+import { useTheme } from '@react-navigation/native';
+import React, { useState } from 'react';
+import { Modal, StyleSheet, TouchableOpacity, View } from 'react-native';
+import { SafeAreaView } from 'react-native-safe-area-context';
+import { typography } from '../../config/typography';
+import FilterButton from '../buttons/FilterButton';
+import PrimaryButton from '../buttons/PrimaryButton';
+import Row from '../common/Row';
+import Space from '../common/Space';
+import StyledText from '../common/StyledText';
+
+type Filters = {
+  level: Array<string>;
+  topic: Array<string>;
+  room: Array<string>;
+  sessionType: Array<string>;
+};
+
+type FilterModalProps = {
+  visible: boolean;
+  onClose: () => void;
+  onFilter: (filters: Filters) => void;
 };
 
+const FilterModal = ({ visible, onClose, onFilter }: FilterModalProps) => {
+  const { colors } = useTheme();
+  const { primary } = typography;
+
+  const [selectedLevel, setSelectedLevel] = useState<Array<string>>([]);
+  const [selectedTopic, setSelectedTopic] = useState<Array<string>>([]);
+  const [selectedRoom, setSelectedRoom] = useState<Array<string>>([]);
+  const [selectedSessionType, setSelectedSessionType] = useState<Array<string>>([]);
```

Comment:
The filter is much better now and allows for multi selection 👍 

However, we are not making good use of TypeScript here to prevent us from breaking the filters since the `string` type is too broad.

For example, right now it is possible to ship code like this:

```tsx
onPress={() => setSelectedLevel(addFilter('Hello', selectedLevel))}
```

where we pass an invalid filter to the `addFilter` function and this will result in a broken UI but TypeScript does not report this because we have not given it enough information. We should only allow valid values.

Can you make this change to tell TypeScript what the valid values are so that such errors can be reported to us:

```tsx
type LevelFilter = 'Beginner' | 'Intermediate' | 'Advanced';
type TopicFilter = 'UI/UX Design' | 'Backend' | 'APIs';
type RoomFilter = 'Room 1' | 'Room 2' | 'Room 3';
type SessionTypeFilter = 'Keynote' | 'CodeLab' | 'Session' | 'Lightning Talk' | 'Panel Discussion';

type Filters = {
  level: Array<LevelFilter>;
  topic: Array<TopicFilter>;
  room: Array<RoomFilter>;
  sessionType: Array<SessionTypeFilter>;
};

// <skipped>

const [selectedLevel, setSelectedLevel] = useState<Filters['level']>([]);
const [selectedTopic, setSelectedTopic] = useState<Filters['topic']>([]);
const [selectedRoom, setSelectedRoom] = useState<Filters['room']>([]);
const [selectedSessionType, setSelectedSessionType] = useState<Filters['sessionType']>([]);

// <skipped>

const addFilter = <TFilter extends string>(filter: TFilter, filterArray: Array<TFilter>) => {
  // <skipped>
};
```

Now TypeScript correctly reports the invalid usage, and also provides us with autocompletion:

```tsx
onPress={() => setSelectedLevel(addFilter('Hello', selectedLevel))}
//                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//                              Type '"Hello"' is not assignable to type 'LevelFilter'.
```

Reply:
this is interesting. I couldn't have thought of it